### ISSUE TITLE
False positive: Unnecessary type casting

### ISSUE THREAD

**Issue by gharlan** [kalessil/phpinspectionsea#490]
Examples: https://3v4l.org/rp4qi9

The plugin says the int casting is not necessary, but as you can see, result without casting is different.

**Comment by kalessil** [comment]
The inspection relies on types information, so the original code or it's synthetic example is needed for root cause analysis. Would you please share one of them?

**Comment by gharlan** [comment]
It is reproducible with the code from my link above.
The return type of the function is float/string.

![screenshot 2017-08-28 16 32 34](https://user-images.githubusercontent.com/330436/29778121-8d7bc012-8c0e-11e7-9da2-2bb6b230f834.png)

**Comment by gharlan** [comment]
Hmm, it seems to be a phpstorm bug, a wrong type is resolved:

![screenshot 2017-08-28 16 39 07](https://user-images.githubusercontent.com/330436/29778379-7cf6d0fa-8c0f-11e7-9a56-acb6e7919c85.png)

Should be float, not int.

**Comment by gharlan** [comment]
Found an existing issue: https://youtrack.jetbrains.com/issue/WI-37466

**Comment by gharlan** [comment]
So should we close it here?

**Comment by kalessil** [comment]
Oh, that;s mine issue, not yet - we have a workaround for the issue, I'll check if we can extend it to this case.

**Comment by rentalhost** [comment]
I think that to make it easier, just ignore this inspection for group expressions like that, and just wait for the fix from PS team.

**Comment by kalessil** [comment]
Fixed (I pushed a fresh binary as well), closing the issue.

**Comment by kalessil** [comment]
BTW, @gharlan thank you for reporting =)

**Comment by garak** [comment]
I'm getting the same problem with latest versions of phpstorm (2019.1) and this plugin (v3.0.13)
My code:
```php
$foo = 2.077;
$bar = (int) ($foo * 100);
```
So `$bar` value is `207.7` but inspection is reporting a false positive "This type casting is not necessary, as the argument is of needed type"

**Comment by kalessil** [comment]
@garak: would you please create a new issue for your case (is the code fragment in function/method or in "flat script")?

**Comment by garak** [comment]
See #1308

**Comment by KriqN** [comment]
I'm not sure if it is a real problem, or I just don't understand something, thats why I write a comment here instead of creating new issue.
PhpStorm 2020.3
Plugin 4.0.6
PHP 7.3

```
$price = '250.72';
$priceInt = (int) ($price * 100);
```
So `$price * 100` returns float, but inspection is reporting "This type casting is not necessary, as the argument is of needed type".

![EA float int](https://user-images.githubusercontent.com/63552810/101750557-d7edd600-3ad7-11eb-8769-63348697cf47.png)
